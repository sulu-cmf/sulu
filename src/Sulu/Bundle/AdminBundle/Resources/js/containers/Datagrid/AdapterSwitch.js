// @flow
import React from 'react';
import ButtonGroup from '../../components/ButtonGroup';
import Button from '../../components/Button';
import Icon from '../../components/Icon';
import datagridAdapterRegistry from './registries/DatagridAdapterRegistry';

type Props = {
    adapters: Array<string>,
    currentAdapter: string,
    onAdapterChange: (adapter: string) => void,
};

export default class AdapterSwitch extends React.PureComponent<Props> {
    handleAdapterChange = (adapter: ?string) => {
        if (!adapter || this.props.currentAdapter === adapter) {
            return;
        }

        this.props.onAdapterChange(adapter);
    };

    render() {
        const {
            currentAdapter,
            adapters,
        } = this.props;

        if (adapters.length < 2) {
            return null;
        }

        return (
            <ButtonGroup>
                {adapters.map((adapter, index) => {
                    const Adapter = datagridAdapterRegistry.get(adapter);

                    return (
                        <Button
                            active={adapter === currentAdapter}
                            key={index}
                            onClick={this.handleAdapterChange}
                            value={adapter}
                        >
                            <Icon name={Adapter.icon} />
                        </Button>
                    );
                })}
            </ButtonGroup>
        );
    }
}
