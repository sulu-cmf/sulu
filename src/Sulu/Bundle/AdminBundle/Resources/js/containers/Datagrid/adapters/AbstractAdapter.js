// @flow
import React from 'react';
import type {LoadingStrategyInterface, DatagridAdapterProps, StructureStrategyInterface} from '../types';

export default class AbstractAdapter extends React.Component<DatagridAdapterProps> {
    static LoadingStrategy: Class<LoadingStrategyInterface>;

    static StructureStrategy: Class<StructureStrategyInterface>;
}
