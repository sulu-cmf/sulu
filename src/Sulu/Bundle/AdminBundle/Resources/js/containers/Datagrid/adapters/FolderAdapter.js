// @flow
import {observer} from 'mobx-react';
import React from 'react';
import FolderList from '../../../components/FolderList';
import type {DatagridAdapterProps} from '../types';
import {translate} from '../../../services/Translator';

@observer
export default class FolderAdapter extends React.Component<DatagridAdapterProps> {
    static getInfoText(item: Object) {
        const label = (item.objectCount === 1)
            ? translate('sulu_admin.object')
            : translate('sulu_admin.objects');

        return `${item.objectCount} ${label}`;
    }

    render() {
        const {
            data,
            onItemClick,
        } = this.props;

        return (
            <FolderList onFolderClick={onItemClick}>
                {data.map((item: Object) => (
                    // TODO: Don't access properties like "title" directly.
                    <FolderList.Folder
                        key={item.id}
                        id={item.id}
                        title={item.title}
                        info={FolderAdapter.getInfoText(item)}
                    />
                ))}
            </FolderList>
        );
    }
}
