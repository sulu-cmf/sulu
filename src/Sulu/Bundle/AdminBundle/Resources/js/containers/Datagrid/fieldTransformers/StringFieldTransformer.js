// @flow
import type {FieldTransformer} from '../types';

export default class StringFieldTransformer implements FieldTransformer {
    transform(value: *): * {
        return value;
    }
}
