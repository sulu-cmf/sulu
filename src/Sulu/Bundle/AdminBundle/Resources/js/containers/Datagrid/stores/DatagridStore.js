// @flow
import {action, autorun, intercept, observable, computed} from 'mobx';
import type {IValueWillChange} from 'mobx'; // eslint-disable-line import/named
import type {ObservableOptions} from '../types';
import ResourceRequester from '../../../services/ResourceRequester';
import metadataStore from './MetadataStore';

export default class DatagridStore {
    @observable pageCount: number = 0;
    @observable data: Array<Object> = [];
    @observable selections: Array<string | number> = [];
    @observable loading: boolean = true;
    @observable loadingStrategy: ?string;
    disposer: () => void;
    resourceKey: string;
    options: Object;
    observableOptions: ObservableOptions;
    localeInterceptionDisposer: () => void;

    constructor(
        resourceKey: string,
        observableOptions: ObservableOptions,
        options: Object = {}
    ) {
        this.resourceKey = resourceKey;
        this.observableOptions = observableOptions;
        this.options = options;
        this.disposer = autorun(this.sendRequest);
    }

    @computed get initialized(): boolean {
        return !!this.loadingStrategy;
    }

    @action init = (loadingStrategy: string) => {
        this.updateLoadingStrategy(loadingStrategy);
    };

    @action updateLoadingStrategy = (loadingStrategy: string) => {
        if (this.loadingStrategy === loadingStrategy) {
            return;
        }

        this.data = [];
        this.pageCount = 0;
        this.setPage(1);
        this.loadingStrategy = loadingStrategy;

        if (this.localeInterceptionDisposer) {
            this.localeInterceptionDisposer();
        }

        if ('infiniteScroll' === this.loadingStrategy && this.observableOptions.locale) {
            this.localeInterceptionDisposer = intercept(this.observableOptions.locale, '', this.handleLocaleChanges);
        }
    };

    handleLocaleChanges = (change: IValueWillChange<number>) => {
        if (this.observableOptions.locale !== change.newValue) {
            this.data = [];
            this.observableOptions.page.set(1);

            return change;
        }
    };

    getSchema() {
        return metadataStore.getSchema(this.resourceKey);
    }

    sendRequest = () => {
        if (!this.initialized) {
            return;
        }

        const page = this.getPage();
        const loadingStrategy = this.loadingStrategy;

        if (!page) {
            return;
        }

        const observableOptions = {};
        observableOptions.page = page;

        if (this.observableOptions.locale) {
            observableOptions.locale = this.observableOptions.locale.get();
        }

        this.setLoading(true);

        ResourceRequester.getList(this.resourceKey, {
            ...observableOptions,
            ...this.options,
        }).then(action((response) => {
            this.handleResponse(response, loadingStrategy);
        }));
    };

    handleResponse = (response: Object, loadingStrategy: ?string) => {
        const data = response._embedded[this.resourceKey];

        if ('infiniteScroll' === loadingStrategy) {
            this.data = [...this.data, ...data];
        } else {
            this.data = data;
        }

        this.pageCount = response.pages;
        this.setLoading(false);
    };

    @action setLoading(loading: boolean) {
        this.loading = loading;
    }

    getPage(): ?number {
        const page = parseInt(this.observableOptions.page.get());
        if (!page) {
            return undefined;
        }

        return page;
    }

    @action setPage(page: number) {
        this.observableOptions.page.set(page);
    }

    @action select(id: string | number) {
        if (this.selections.includes(id)) {
            return;
        }

        this.selections.push(id);
    }

    @action selectEntirePage() {
        this.data.forEach((item) => {
            this.select(item.id);
        });
    }

    @action deselect(id: string | number) {
        const index = this.selections.indexOf(id);
        if (index === -1) {
            return;
        }

        this.selections.splice(index, 1);
    }

    @action deselectEntirePage() {
        this.data.forEach((item) => {
            this.deselect(item.id);
        });
    }

    @action clearSelection() {
        this.selections = [];
    }

    destroy() {
        this.disposer();

        if (this.localeInterceptionDisposer) {
            this.localeInterceptionDisposer();
        }
    }
}
