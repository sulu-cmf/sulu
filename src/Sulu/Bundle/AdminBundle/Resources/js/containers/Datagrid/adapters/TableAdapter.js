// @flow
import {observer} from 'mobx-react';
import React from 'react';
import Pagination from '../../../components/Pagination';
import Table from '../../../components/Table';
import {translate} from '../../../utils/Translator';
import PaginatedLoadingStrategy from '../loadingStrategies/PaginatedLoadingStrategy';
import FlatStructureStrategy from '../structureStrategies/FlatStructureStrategy';
import AbstractAdapter from './AbstractAdapter';

@observer
export default class TableAdapter extends AbstractAdapter {
    static LoadingStrategy = PaginatedLoadingStrategy;

    static StructureStrategy = FlatStructureStrategy;

    static icon = 'su-align-justify';

    static defaultProps = {
        data: [],
    };

    renderCells(item: Object, schemaKeys: Array<string>) {
        return schemaKeys.map((schemaKey) => {
            // TODO: Remove this when a datafield mapping is built
            if (typeof item[schemaKey] === 'object') {
                return <Table.Cell key={item.id + schemaKey}>Object!</Table.Cell>;
            }

            return (
                <Table.Cell key={item.id + schemaKey}>{item[schemaKey]}</Table.Cell>
            );
        });
    }

    renderHeaderCells(schema: Object, schemaKeys: Array<string>) {
        return schemaKeys.map((schemaKey) => {
            const title = schema[schemaKey] && schema[schemaKey].title ? translate(schema[schemaKey].title) : schemaKey;

            return(
                <Table.HeaderCell key={schemaKey}>
                    {title}
                </Table.HeaderCell>
            );
        });
    }

    render() {
        const {
            data,
            loading,
            onItemClick,
            onAllSelectionChange,
            onItemSelectionChange,
            onPageChange,
            page,
            pageCount,
            schema,
            selections,
        } = this.props;
        const schemaKeys = Object.keys(schema);
        const buttons = [];

        if (onItemClick) {
            buttons.push({
                icon: 'su-pen',
                onClick: (rowId) => onItemClick(rowId),
            });
        }

        return (
            <Pagination
                total={pageCount}
                current={page}
                loading={loading}
                onChange={onPageChange}
            >
                <Table
                    buttons={buttons}
                    selectMode={onItemSelectionChange ? 'multiple' : undefined}
                    onRowSelectionChange={onItemSelectionChange}
                    onAllSelectionChange={onAllSelectionChange}
                >
                    <Table.Header>
                        {this.renderHeaderCells(schema, schemaKeys)}
                    </Table.Header>
                    <Table.Body>
                        {data.map((item) => (
                            <Table.Row key={item.id} id={item.id} selected={selections.includes(item.id)}>
                                {this.renderCells(item, schemaKeys)}
                            </Table.Row>
                        ))}
                    </Table.Body>
                </Table>
            </Pagination>
        );
    }
}
