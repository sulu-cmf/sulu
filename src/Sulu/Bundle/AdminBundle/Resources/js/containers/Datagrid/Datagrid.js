// @flow
import {observer} from 'mobx-react';
import {observable, action, computed} from 'mobx';
import React, {Fragment} from 'react';
import type {Node} from 'react';
import equal from 'fast-deep-equal';
import Dialog from '../../components/Dialog';
import {translate} from '../../utils/Translator';
import type {SortOrder} from './types';
import DatagridStore from './stores/DatagridStore';
import datagridAdapterRegistry from './registries/DatagridAdapterRegistry';
import AbstractAdapter from './adapters/AbstractAdapter';
import AdapterSwitch from './AdapterSwitch';
import Search from './Search';
import datagridStyles from './datagrid.scss';

type Props = {|
    adapters: Array<string>,
    deletable: boolean,
    disabledIds: Array<string | number>,
    header?: Node,
    onItemClick?: (itemId: string | number) => void,
    onAddClick?: (id: string | number) => void,
    selectable: boolean,
    searchable: boolean,
    store: DatagridStore,
|};

@observer
export default class Datagrid extends React.Component<Props> {
    static defaultProps = {
        deletable: true,
        disabledIds: [],
        selectable: true,
        searchable: true,
    };

    @observable currentAdapterKey: string;
    @observable showDeleteDialog: boolean = false;
    @observable deleting: boolean = false;
    deleteId: ?string | number;

    @computed get currentAdapter(): typeof AbstractAdapter {
        return datagridAdapterRegistry.get(this.currentAdapterKey);
    }

    @computed get currentAdapterOptions(): typeof AbstractAdapter {
        return datagridAdapterRegistry.getOptions(this.currentAdapterKey);
    }

    constructor(props: Props) {
        super(props);

        this.validateAdapters();
    }

    componentDidUpdate(prevProps: Props) {
        if (!equal(this.props.adapters, prevProps.adapters)) {
            this.validateAdapters();
        }

        if (this.props.store !== prevProps.store) {
            this.props.store.updateStrategies(
                new this.currentAdapter.LoadingStrategy(),
                new this.currentAdapter.StructureStrategy()
            );
        }
    }

    validateAdapters() {
        const {adapters} = this.props;

        adapters.forEach((adapterName) => {
            if (!datagridAdapterRegistry.has(adapterName)) {
                throw new Error(
                    'DatagridAdapter with the name "' + adapterName + '" does not exist.' +
                    'Did you forget to add it to the "datagridAdapterRegistry"?'
                );
            }
        });

        if (!this.currentAdapterKey) {
            this.setCurrentAdapterKey(this.props.adapters[0]);
        }
    }

    @action setCurrentAdapterKey = (adapter: string) => {
        this.currentAdapterKey = adapter;
        this.props.store.updateStrategies(
            new this.currentAdapter.LoadingStrategy(),
            new this.currentAdapter.StructureStrategy()
        );
    };

    @action handleDeleteClick = (id: string | number) => {
        this.deleteId = id;
        this.showDeleteDialog = true;
    };

    @action handleDeleteDialogConfirmClick = () => {
        if (!this.deleteId) {
            throw new Error('The id for deletion was not set. This should not happen, and is like caused by a bug.');
        }

        this.deleting = true;
        this.props.store.delete(this.deleteId).then(action(() => {
            this.deleteId = undefined;
            this.deleting = false;
            this.showDeleteDialog = false;
        }));
    };

    @action handleDeleteDialogCancelClick = () => {
        this.showDeleteDialog = false;
    };

    handlePageChange = (page: number) => {
        this.props.store.setPage(page);
    };

    handleSort = (column: string, order: SortOrder) => {
        this.props.store.sort(column, order);
    };

    handleSearch = (search: ?string) => {
        this.props.store.search(search);
    };

    handleItemSelectionChange = (id: string | number, selected?: boolean) => {
        const {store} = this.props;
        const row = store.findById(id);

        if (!row) {
            return;
        }

        selected ? store.select(row) : store.deselect(row);
    };

    handleAllSelectionChange = (selected?: boolean) => {
        const {store} = this.props;
        selected ? store.selectVisibleItems() : store.deselectVisibleItems();
    };

    handleAdapterChange = (adapter: string) => {
        this.props.store.setActive(undefined); // TODO keep active and expand correctly
        this.setCurrentAdapterKey(adapter);
    };

    handleItemActivation = (id: string | number) => {
        this.props.store.activate(id);
    };

    handleItemDeactivation = (id: string | number) => {
        this.props.store.deactivate(id);
    };

    render() {
        const {
            adapters,
            deletable,
            disabledIds,
            header,
            onItemClick,
            onAddClick,
            searchable,
            selectable,
            store,
        } = this.props;
        const Adapter = this.currentAdapter;

        return (
            <Fragment>
                {(header || searchable || adapters.length > 1) &&
                    <div className={datagridStyles.headerContainer}>
                        {header}
                        <div className={datagridStyles.toolbar}>
                            {searchable &&
                                <Search onSearch={this.handleSearch} value={store.searchTerm.get()} />
                            }
                            <AdapterSwitch
                                adapters={adapters}
                                currentAdapter={this.currentAdapterKey}
                                onAdapterChange={this.handleAdapterChange}
                            />
                        </div>
                    </div>
                }
                <div className={datagridStyles.datagrid}>
                    <Adapter
                        active={store.active.get()}
                        activeItems={store.activeItems}
                        data={store.data}
                        disabledIds={disabledIds}
                        loading={store.loading}
                        onAddClick={onAddClick}
                        onAllSelectionChange={selectable ? this.handleAllSelectionChange : undefined}
                        onDeleteClick={deletable ? this.handleDeleteClick : undefined}
                        onItemActivation={this.handleItemActivation}
                        onItemDeactivation={this.handleItemDeactivation}
                        onItemClick={onItemClick}
                        onItemSelectionChange={selectable ? this.handleItemSelectionChange : undefined}
                        onPageChange={this.handlePageChange}
                        onSort={this.handleSort}
                        options={this.currentAdapterOptions}
                        page={store.getPage()}
                        pageCount={store.pageCount}
                        schema={store.schema}
                        sortColumn={store.sortColumn.get()}
                        sortOrder={store.sortOrder.get()}
                        selections={store.selectionIds}
                    />
                </div>
                <Dialog
                    confirmLoading={this.deleting}
                    cancelText={translate('sulu_admin.cancel')}
                    confirmText={translate('sulu_admin.ok')}
                    onCancel={this.handleDeleteDialogCancelClick}
                    onConfirm={this.handleDeleteDialogConfirmClick}
                    open={this.showDeleteDialog}
                    title={translate('sulu_admin.delete_warning_title')}
                >
                    {translate('sulu_admin.delete_warning_text')}
                </Dialog>
            </Fragment>
        );
    }
}
