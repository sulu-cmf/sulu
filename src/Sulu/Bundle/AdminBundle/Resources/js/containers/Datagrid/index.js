// @flow
import Datagrid from './Datagrid';
import DatagridStore from './stores/DatagridStore';
import datagridAdapterStore from './stores/DatagridAdapterStore';
import TableAdapter from './adapters/TableAdapter';
import FolderAdapter from './adapters/FolderAdapter';
import type {DatagridAdapter, DatagridAdapterProps} from './types';

export default Datagrid;
export {
    DatagridStore,
    datagridAdapterStore,
    TableAdapter,
    FolderAdapter,
};
export type {
    DatagridAdapter,
    DatagridAdapterProps,
};
