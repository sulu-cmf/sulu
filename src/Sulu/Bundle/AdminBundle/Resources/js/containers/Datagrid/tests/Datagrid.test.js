// @flow
import {mount, shallow} from 'enzyme';
import React from 'react';
import {observable} from 'mobx';
import Datagrid from '../Datagrid';
import DatagridStore from '../stores/DatagridStore';
import datagridAdapterRegistry from '../registries/DatagridAdapterRegistry';
import datagridFieldTransformerRegistry from '../registries/DatagridFieldTransformerRegistry';
import AbstractAdapter from '../adapters/AbstractAdapter';
import TableAdapter from '../adapters/TableAdapter';
import FolderAdapter from '../adapters/FolderAdapter';
import StringFieldTransformer from '../fieldTransformers/StringFieldTransformer';

jest.mock('../stores/DatagridStore', () => jest.fn(function() {
    this.setPage = jest.fn();
    this.setActive = jest.fn();
    this.sort = jest.fn();
    this.sortColumn = {
        get: jest.fn(),
    };
    this.sortOrder = {
        get: jest.fn(),
    };
    this.updateStrategies = jest.fn();
    this.getPage = jest.fn().mockReturnValue(4);
    this.pageCount = 7;
    this.selections = [];
    this.selectionIds = [];
    this.loading = false;
    this.schema = {
        title: {
            type: 'string',
            visibility: 'yes',
            label: 'Title',
        },
    };
    this.findById = jest.fn();
    this.select = jest.fn();
    this.deselect = jest.fn();
    this.selectEntirePage = jest.fn();
    this.deselectEntirePage = jest.fn();
    this.updateLoadingStrategy = jest.fn();
    this.structureStrategy = {
        data: [
            {
                title: 'value',
                id: 1,
            },
        ],
    };
    this.data = this.structureStrategy.data;
}));

jest.mock('../registries/DatagridAdapterRegistry', () => ({
    add: jest.fn(),
    get: jest.fn(),
    has: jest.fn(),
}));

jest.mock('../registries/DatagridFieldTransformerRegistry', () => ({
    add: jest.fn(),
    get: jest.fn(),
    has: jest.fn(),
}));

jest.mock('../../../utils/Translator', () => ({
    translate: function(key) {
        switch (key) {
            case 'sulu_admin.page':
                return 'Page';
            case 'sulu_admin.of':
                return 'of';
        }
    },
}));

class LoadingStrategy {
    load = jest.fn();
    destroy = jest.fn();
    initialize = jest.fn();
    reset = jest.fn();
}

class StructureStrategy {
    data: Array<Object>;

    clear = jest.fn();
    getData = jest.fn();
    findById = jest.fn();
    enhanceItem = jest.fn();
}

class TestAdapter extends AbstractAdapter {
    static LoadingStrategy = LoadingStrategy;

    static StructureStrategy = StructureStrategy;

    static icon = 'su-th-large';

    render() {
        return (
            <div>Test Adapter</div>
        );
    }
}

beforeEach(() => {
    datagridAdapterRegistry.has.mockReturnValue(true);
    datagridAdapterRegistry.get.mockReturnValue(TestAdapter);

    datagridFieldTransformerRegistry.get.mockReturnValue(new StringFieldTransformer());
});

test('Render TableAdapter with correct values', () => {
    datagridAdapterRegistry.get.mockReturnValue(TableAdapter);

    const datagridStore = new DatagridStore('test', {page: observable.box(1)});
    datagridStore.active = 3;
    datagridStore.selectionIds.push(1, 3);
    const editClickSpy = jest.fn();

    const datagrid = shallow(<Datagrid adapters={['table']} store={datagridStore} onItemClick={editClickSpy} />);
    const tableAdapter = datagrid.find('TableAdapter');

    expect(tableAdapter.prop('data')).toEqual([{'id': 1, 'title': 'value'}]);
    expect(tableAdapter.prop('active')).toEqual(3);
    expect(tableAdapter.prop('selections')).toEqual([1, 3]);
    expect(tableAdapter.prop('schema')).toEqual({
        title: {
            type: 'string',
            visibility: 'yes',
            label: 'Title',
        },
    });
    expect(tableAdapter.prop('onItemClick')).toBe(editClickSpy);
    expect(tableAdapter.prop('onItemSelectionChange')).toBeInstanceOf(Function);
    expect(tableAdapter.prop('onAllSelectionChange')).toBeInstanceOf(Function);
});

test('Render the adapter in non-selectable mode', () => {
    const datagridStore = new DatagridStore('test', {page: observable.box(1)});
    const datagrid = shallow(<Datagrid adapters={['test']} selectable={false} store={datagridStore} />);

    expect(datagrid.find('TestAdapter').prop('onItemSelectionChange')).toEqual(undefined);
    expect(datagrid.find('TestAdapter').prop('onAllSelectionChange')).toEqual(undefined);
});

test('Pass the ids to be disabled to the adapter', () => {
    const disabledIds = [1, 3];
    const datagridStore = new DatagridStore('test', {page: observable.box(1)});
    const datagrid = shallow(<Datagrid adapters={['test']} disabledIds={disabledIds} store={datagridStore} />);

    expect(datagrid.find('TestAdapter').prop('disabledIds')).toBe(disabledIds);
});

test('Pass sortColumn and sortOrder to adapter', () => {
    const datagridStore = new DatagridStore('test', {page: observable.box(1)});
    datagridStore.sortColumn.get.mockReturnValue('title');
    datagridStore.sortOrder.get.mockReturnValue('asc');
    const datagrid = shallow(<Datagrid adapters={['test']} store={datagridStore} />);

    expect(datagrid.find('TestAdapter').props()).toEqual(expect.objectContaining({
        sortColumn: 'title',
        sortOrder: 'asc',
    }));
});

test('Selecting and deselecting items should update store', () => {
    datagridAdapterRegistry.get.mockReturnValue(TableAdapter);
    const datagridStore = new DatagridStore('test', {page: observable.box(1)});
    datagridStore.structureStrategy.data.splice(0, datagridStore.structureStrategy.data.length);
    datagridStore.structureStrategy.data.push(
        {id: 1},
        {id: 2},
        {id: 3}
    );

    datagridStore.findById.mockReturnValueOnce({id: 1}).mockReturnValueOnce({id: 2}).mockReturnValueOnce({id: 1});

    const datagrid = mount(<Datagrid adapters={['table']} store={datagridStore} />);

    const checkboxes = datagrid.find('input[type="checkbox"]');
    // TODO setting checked explicitly should not be necessary, see https://github.com/airbnb/enzyme/issues/1114
    checkboxes.at(1).getDOMNode().checked = true;
    checkboxes.at(2).getDOMNode().checked = true;
    checkboxes.at(1).simulate('change', {currentTarget: {checked: true}});
    expect(datagridStore.findById).toBeCalledWith(1);
    expect(datagridStore.select).toBeCalledWith({id: 1});
    checkboxes.at(2).simulate('change', {currentTarget: {checked: true}});
    expect(datagridStore.findById).toBeCalledWith(2);
    expect(datagridStore.select).toBeCalledWith({id: 2});
    checkboxes.at(1).simulate('change', {currentTarget: {checked: false}});
    expect(datagridStore.findById).toBeCalledWith(1);
    expect(datagridStore.deselect).toBeCalledWith({id: 1});
});

test('Selecting and unselecting all items on current page should update store', () => {
    datagridAdapterRegistry.get.mockReturnValue(TableAdapter);
    const datagridStore = new DatagridStore('test', {page: observable.box(1)});
    datagridStore.structureStrategy.data = [
        {id: 1},
        {id: 2},
        {id: 3},
    ];
    const datagrid = mount(<Datagrid adapters={['table']} store={datagridStore} />);

    const headerCheckbox = datagrid.find('input[type="checkbox"]').at(0);
    // TODO setting checked explicitly should not be necessary, see https://github.com/airbnb/enzyme/issues/1114
    headerCheckbox.getDOMNode().checked = true;
    headerCheckbox.simulate('change', {currentTarget: {checked: true}});
    expect(datagridStore.selectEntirePage).toBeCalledWith();
    headerCheckbox.simulate('change', {currentTarget: {checked: false}});
    expect(datagridStore.deselectEntirePage).toBeCalledWith();
});

test('Clicking a header cell should sort the table', () => {
    datagridAdapterRegistry.get.mockReturnValue(TableAdapter);
    const datagridStore = new DatagridStore('test', {page: observable.box(1)});
    datagridStore.structureStrategy.data = [
        {id: 1},
        {id: 2},
        {id: 3},
    ];
    const datagrid = mount(<Datagrid adapters={['table']} store={datagridStore} />);

    const headerCell = datagrid.find('th button').at(0);
    headerCell.simulate('click');
    expect(datagridStore.sort).toBeCalledWith('title', 'asc');
});

test('Switching the adapter should render the correct adapter', () => {
    const datagridStore = new DatagridStore('test', {page: observable.box(1)});

    datagridAdapterRegistry.get.mockImplementation((adapter) => {
        switch (adapter) {
            case 'table':
                return TableAdapter;
            case 'folder':
                return FolderAdapter;
        }
    });
    const datagrid = mount(<Datagrid adapters={['table', 'folder']} store={datagridStore} />);

    expect(datagrid.find('AdapterSwitch').length).toBe(1);
    expect(datagrid.find('TableAdapter').length).toBe(1);

    datagrid.find('AdapterSwitch Button').at(1).simulate('click');
    expect(datagrid.find('TableAdapter').length).toBe(0);
    expect(datagrid.find('FolderAdapter').length).toBe(1);
});

test('DatagridStore should be initialized correctly on init and update', () => {
    const datagridStore = new DatagridStore('test', {page: observable.box(1)});
    datagridStore.updateStrategies = jest.fn();

    datagridAdapterRegistry.get.mockImplementation((adapter) => {
        switch (adapter) {
            case 'table':
                return TableAdapter;
            case 'folder':
                return FolderAdapter;
        }
    });
    mount(<Datagrid adapters={['table', 'folder']} store={datagridStore} />);
    expect(datagridStore.updateStrategies)
        .toBeCalledWith(expect.any(TableAdapter.LoadingStrategy), expect.any(TableAdapter.StructureStrategy));
});

test('DatagridStore should be updated with current active element', () => {
    datagridAdapterRegistry.get.mockReturnValue(class TestAdapter extends AbstractAdapter {
        static LoadingStrategy = class {
            paginationAdapter = undefined;
            load = jest.fn();
            destroy = jest.fn();
            initialize = jest.fn();
            reset = jest.fn();
        };

        static StructureStrategy = class {
            data = [];
            clear = jest.fn();
            getData = jest.fn();
            findById = jest.fn();
            enhanceItem = jest.fn();
        };

        static icon = 'su-th-large';

        constructor(props: *) {
            super(props);

            const {onItemActivation} = this.props;
            if (onItemActivation) {
                onItemActivation('some-uuid');
            }
        }

        render() {
            return null;
        }
    });
    const datagridStore = new DatagridStore('test', {page: observable.box(1)});
    expect(datagridStore.active).toBe(undefined);
    mount(<Datagrid adapters={['test']} store={datagridStore} />);

    expect(datagridStore.setActive).toBeCalledWith('some-uuid');
});
