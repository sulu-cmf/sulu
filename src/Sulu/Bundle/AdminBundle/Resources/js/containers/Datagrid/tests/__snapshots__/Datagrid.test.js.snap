// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render the adapter in non-searchable mode 1`] = `
Array [
  <div
    class="headerContainer"
  >
    <h1>
      Title
    </h1>
    <div
      class="toolbar"
    />
  </div>,
  <div
    class="datagrid"
  >
    <div>
      Test Adapter
    </div>
  </div>,
  <div />,
]
`;
